### PR TITLE
Add TestResult to MultipleAssertException

### LINKED ISSUE #2695: MultipleAssertException doesn't provide proper details on failures
When using `Assert.Multiple` and assertion failures happen inside the lambda, a `MultipleAssertException` is thrown by the framework.

**This exception has absolutely no details on the actual failures that happened**, which drastically impacts the debugging experience: users have to wait for the run to finish and then inspect the Test Explorer window to see what errors have occurred.

This is where it is being thrown currently:
`https://github.com/nunit/nunit/blob/efd115b26c8f5d4ebe6d52f5108097ceb0c77ffb/src/NUnitFramework/framework/Assert.cs#L340`
```
            if (context.MultipleAssertLevel == 0 && context.CurrentResult.PendingFailures > 0)
                throw new MultipleAssertException();
```

This results in the following message:
> "One or more failures in Multiple Assert block:"

This gives no hint to what happened and leaves the developer wondering how to see the actual failures. I think this is a really bad practice with exceptions in general: they should provide as much detail as humanly possible.

I'm aware that the display on Test Explorer has full details, but NUnit shouldn't assume that all developers _run_ tests instead of _debugging_ them.

I propose the following:
1. **That failures be integrated as inner exceptions**
This would be similar to how an `AggregateException` works: have all failures as child exceptions so that people debugging the application can see what errors occurred. Maybe using an `AggregateException` would be ideal, but I heard you had some issues with that because the type doesn't exist in all versions of the framework. Instead of just inspecting the `PendingFailures` counter like above, we could convert any `AssertionResult` into `AssertionException`s and add them to the parent `MultipleAssertException` before throwing it. I'd even suggest ditching the `AssertionResult` model itself in favor of directly handling `AssertionException` objects, but that's probably a bigger discussion that I don't want to get into right now.

2. **Create an user-friedly `.ToString` implementation on `MultipleAssertException`**
In combination with # 1, the exception's `ToString` should provide a summary of failures. That would allow developers to see what errors were reported easily, and only have to inspect the inner exceptions in case further details are required. Maybe ommiting the stacktrace for each error on this summary would be a good idea, for example.

This lack of debugging support had a very significant impact when migrating to `Assert.Multiple` on our team: we used to have a custom method to perform this task which reported a much cleaner error in debug mode showing all assertions that failed, and when we switched to `Assert.Multiple` there was justified backlash.

### PR DESCRIPTION
**Pull request by CharliePoole**
This fixes #2695 by including the test result and combined failure/warning messages in the MultipleAssert exception for use in debugging. In addition, the same thing is done for AssertionException.

Additional tests were added to highlight problems with warning messages but the tests are currently ignored since they don't pass. See issue #2699

It's clear that more work is needed in the recording of multiple results and in generally cleaning up the `TestResult` class. This PR should serve as a good start.

### CODE REVIEW

File: src/NUnitFramework/framework/Exceptions/MultipleAssertException.cs [hunk 1]
```diff
@@ -64,5 +76,10 @@ public override ResultState ResultState
         {
             get { return ResultState.Failure; }
         }
+
+        /// <summary>
+        /// Gets the ITestResult of this test at the point the exception was thrown
```
Comment (julealgon) [inline src/NUnitFramework/framework/Exceptions/MultipleAssertException.cs]: Missing period at the end.

File: src/NUnitFramework/framework/Exceptions/MultipleAssertException.cs [hunk 2]
```diff
@@ -36,9 +36,21 @@ namespace NUnit.Framework
...
+        public MultipleAssertException(string message) : base(message) { }
+
+        /// <summary>
+        /// Construct based on the TestResult so far. This is the constructor
+        /// used normally, when exiting the multiple assert block with failures.
+        /// Not used internally but provided to facilitate debugging.
+        /// </summary>
+        /// <param name="result">The current result</param>
+        public MultipleAssertException(ITestResult result)
+            : base(result.Message)
```
Comment (julealgon) [inline src/NUnitFramework/framework/Exceptions/MultipleAssertException.cs]: Missing argument validation. If `result` is null, this will throw. You'll probably need some inline static validation to keep using it as an argument to the base. Something like this:
```
: base(EnsureNotNull(result).Message)
```

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs [hunk 3]
```diff
@@ -498,24 +498,15 @@ public void RecordException(Exception ex)
...
-            }
+                SetResult(
+                    ((ResultStateException)ex).ResultState,
+                    ex.Message, 
+                    StackFilter.DefaultFilter.Filter(ex.StackTrace));
 #if !NETSTANDARD1_6
             else if (ex is System.Threading.ThreadAbortException)
-                SetResult(ResultState.Cancelled,
+                SetResult(
+                    ResultState.Cancelled,
```
Comment (julealgon) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: Unrelated formatting change, I'd suggest not performing it as part of this PR to avoid behavior+refactor changes in the same boat.

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs [hunk 4]
```diff
@@ -498,24 +498,15 @@ public void RecordException(Exception ex)
...
-                    : ex.Message;
-
-                string stackTrace = StackFilter.DefaultFilter.Filter(ex.StackTrace);
-
-                SetResult(((ResultStateException)ex).ResultState, message, stackTrace);
-            }
+                SetResult(
+                    ((ResultStateException)ex).ResultState,
+                    ex.Message, 
+                    StackFilter.DefaultFilter.Filter(ex.StackTrace));
```
Comment (julealgon) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: I'd suggest keeping the scope (`{}`) for readability (since the if block still has more than one line).

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs [hunk 5]
```diff
@@ -498,24 +498,15 @@ public void RecordException(Exception ex)
...
-                string message = ex is MultipleAssertException
-                    ? CreateLegacyFailureMessage()
-                    : ex.Message;
-
-                string stackTrace = StackFilter.DefaultFilter.Filter(ex.StackTrace);
-
-                SetResult(((ResultStateException)ex).ResultState, message, stackTrace);
-            }
+                SetResult(
+                    ((ResultStateException)ex).ResultState,
```
Comment (julealgon) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: This looks unnecessarily complex. I'd suggest either changing this logic into `as`->`!= null` (to avoid the code duplication), or even better, rely on the new switch with pattern matching.

I know this is probably out of scope for this PR though.

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs [hunk 6]
```diff
@@ -643,6 +635,27 @@ public void RecordAssertion(AssertionStatus status, string message)
...
         }
 
+
+        /// <summary>
+        /// Creates a failure message incorporating failures
+        /// from a Multiple Assert block for use by runners
+        /// that don't know about AssertionResults.
+        /// </summary>
+        /// <returns>Message as a string</returns>
+        public string CreateLegacyFailureMessage()
```
Comment (julealgon) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: Why did we make this `public` now? I don't see any extra usages on the PR.

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs [hunk 7]
```diff
@@ -643,6 +635,27 @@ public void RecordAssertion(AssertionStatus status, string message)
...
+        {
+            var writer = new StringWriter();
```
Comment (julealgon) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: Isn't this `IDisposable`? I'd strongly suggest wrapping in an `using` block.

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs [hunk 8]
```diff
@@ -643,6 +635,27 @@ public void RecordAssertion(AssertionStatus status, string message)
...
+
+            if (AssertionResults.Count > 1)
+                writer.WriteLine("Multiple failures or warnings in test:");
```
Comment (julealgon) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: Looks like we are changing the message here. Is it really related to these changes? I'd suggest not performing such a change as part of this PR.

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs [hunk 9]
```diff
@@ -643,6 +635,27 @@ public void RecordAssertion(AssertionStatus status, string message)
...
+
+            int counter = 0;
+            foreach (var assertion in AssertionResults)
```
Comment (julealgon) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: Consider using the `Select` overload that returns the index instead of maintaining a separate `counter` variable.

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs [hunk 10]
```diff
@@ -643,6 +635,27 @@ public void RecordAssertion(AssertionStatus status, string message)
...
+                writer.WriteLine(string.Format("  {0}) {1}", ++counter, assertion.Message));
```
Comment (julealgon) [inline src/NUnitFramework/framework/Internal/Results/TestResult.cs]: Can we use interpolation here?

File: src/NUnitFramework/tests/Assertions/AssertFailTests.cs [hunk 11]
```diff
@@ -104,13 +104,13 @@ public void AssertFailWorksWithMessageAndArgs()
             Assert.AreEqual(ResultState.Failure, result.ResultState);
             Assert.AreEqual("MESSAGE: 2+2=4", result.Message);
 
-            Assert.AreEqual(result.AssertionResults.Count, 1);
+            Assert.AreEqual(1, result.AssertionResults.Count);
```
Comment (julealgon) [inline src/NUnitFramework/tests/Assertions/AssertFailTests.cs]: These inverted arguments are existing technical debt and not really part of this change, so I'd suggest fixing them in a separate PR.

File: src/NUnitFramework/tests/Assertions/AssertFailTests.cs [hunk 12]
```diff
@@ -104,13 +104,13 @@ public void AssertFailWorksWithMessageAndArgs()
...
             var assertion = result.AssertionResults[0];
             Assert.That(assertion.Status, Is.EqualTo(AssertionStatus.Failed));
             Assert.That(assertion.Message, Is.EqualTo("MESSAGE: 2+2=4"));
         }
 
-        [Test]
+        [Test, Ignore("Currently Fails: Must use Assert.Catch or Assert.Throws")]
```
Comment (julealgon) [inline src/NUnitFramework/tests/Assertions/AssertFailTests.cs]: As mentioned in the issue, I'm personally not fond of this breaking change. Even though NUnit has this goal of separating test termination from the failures, it just results in unintuitive behavior to the end user.

File: src/NUnitFramework/tests/Assertions/AssertFailTests.cs [hunk 13]
```diff
@@ -125,5 +125,29 @@ public void CatchingAssertionExceptionMakesTestPass()
...
+        [Test]
+        public void AssertCatchMakesTestPass()
+        {
+            Assert.Catch(() =>
+            {
+                Assert.Fail("This should not be seen");
+            });
+
+            // Ensure that no spurious info was recorded from the assertion
+            Assert.That(TestExecutionContext.CurrentContext.CurrentResult.AssertionResults.Count, Is.EqualTo(0));
```
Comment (julealgon) [inline src/NUnitFramework/tests/Assertions/AssertFailTests.cs]: Can we use `IsEmpty` instead of checking the `Count` property directly? We have an assertion for empty collections right?

File: src/NUnitFramework/tests/Assertions/AssertFailTests.cs [hunk 14]
```diff
@@ -125,5 +125,29 @@ public void CatchingAssertionExceptionMakesTestPass()
...
+        [Test]
+        public void AssertThrowsMakesTestPass()
+        {
+            Assert.Throws<AssertionException>(() =>
+            {
+                Assert.Fail("This should not be seen");
+            });
+
+            // Ensure that no spurious info was recorded from the assertion
+            Assert.That(TestExecutionContext.CurrentContext.CurrentResult.AssertionResults.Count, Is.EqualTo(0));
```
Comment (julealgon) [inline src/NUnitFramework/tests/Assertions/AssertFailTests.cs]: Same here.

File: src/NUnitFramework/tests/Assertions/AssertWarnTests.cs [hunk 15]
```diff
@@ -51,5 +52,47 @@ public void AssertWarnWorksWithMessageAndArgs()
...
+
+        [Test]
+        public void WarningsAreDisplayedWithFailure()
+        {
+            ITestResult result = TestBuilder.RunTestCase(
+                typeof(WarningFixture),
+                "TwoWarningsAndFailure");
+
+            Assert.That(result.ResultState, Is.EqualTo(ResultState.Failure));
+            Assert.That(result.AssertionResults.Count, Is.EqualTo(3));
```
Comment (julealgon) [inline src/NUnitFramework/tests/Assertions/AssertWarnTests.cs]: Similarly here, don't we have an assertion to check collection counts directly, `Has.Exactly`?

**Comment by CharliePoole** [comment]
@julealgon The description of the PR explains how I broadened the scope, mostly from necessity because simply adding a populated result to the exception resulted in other problems. I'll point out the details inline.

File: src/NUnitFramework/tests/Assertions/AssertFailTests.cs (hunk 12 above)
Comment (CharliePoole) [reply src/NUnitFramework/tests/Assertions/AssertFailTests.cs]: Agreed. OTOH, being able to catch an exception thereby preventing a failure is purely an accident of implementation, even though it's an "obvious" implementation. Note that you have __never__ been above to prevent a multiple assert failure by catching an exception and - of course - you can't prevent warnings that way either.

The only alternative I can see is that asserts not use the current result at all, but just make and update a copy of it, which they then throw. That seems like quite a lot of work and I'd want to apply it to all types of exceptions uniformly if we did it.

File: src/NUnitFramework/tests/Assertions/AssertFailTests.cs (hunk 13 above)
Comment (CharliePoole) [reply src/NUnitFramework/tests/Assertions/AssertFailTests.cs]: I thought of it as "number of assertion results is zero" and wrote it that way. Style, I guess.

File: src/NUnitFramework/tests/Assertions/AssertWarnTests.cs (hunk 15 above)
Comment (CharliePoole) [reply src/NUnitFramework/tests/Assertions/AssertWarnTests.cs]: My personal style is to try to limit myself to the most basic assertions when testing NUnit. Otherwise, I have to worry about what I'm testing in order to avoid using the feature that I am currently testing.

File: src/NUnitFramework/framework/Exceptions/MultipleAssertException.cs (hunk 2 above)
Comment (julealgon) [reply src/NUnitFramework/framework/Exceptions/MultipleAssertException.cs]: > Maybe base(result?.Message) and a Guard in the constructor.

That works but the validation will be misleading as you'd be validating the internal `Message` argument and not the `result` itself. A static validation check in this class can properly indicate that `result` itself is `null`.

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs (hunk 4 above)
Comment (julealgon) [reply src/NUnitFramework/framework/Internal/Results/TestResult.cs]: > Personally, I'm not a fan of that. I think the indentation tells the story well enough. Some people fear errors due to indentation that doesn't actually match the nesting, but I haven't run into problems with it myself.

This is absolutely fine and more of a standardization thing honestly. I know StyleCop has a rule for this but as long as you are not following it in this repo it should be OK to ignore it (after all, the `else` block below suffers from the same issue).

File: src/NUnitFramework/tests/Assertions/AssertFailTests.cs (hunk 13 above)
Comment (julealgon) [reply src/NUnitFramework/tests/Assertions/AssertFailTests.cs]: > I thought of it as "number of assertion results is zero" and wrote it that way. Style, I guess.

On this particular one, I disagree that it is a matter of style. The collection constraints provide better semantic and were created for this exact purpose. Additionally, they also help with reducing the amount of child member access/law of demeter violation (which is already insane over here with 4 levels).

File: src/NUnitFramework/tests/Assertions/AssertWarnTests.cs (hunk 15 above)
Comment (julealgon) [reply src/NUnitFramework/tests/Assertions/AssertWarnTests.cs]: > My personal style is to try to limit myself to the most basic assertions when testing NUnit. Otherwise, I have to worry about what I'm testing in order to avoid using the feature that I am currently testing.

Now this is interesting and I see where you are coming from. Going by that logic, I'd even consider using exclusively the barebones `Assert.IsTrue`.

Comment (CharliePoole) [reply src/NUnitFramework/tests/Assertions/AssertWarnTests.cs]: So it's a balance between good messages and simplicity. To have good messages with IsTrue you have to add a message with each assert.

As a matter of general interest, the nunit 3 syntax came from nunitlite, which I boots trapped with about a half dozen assertions.

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs (hunk 7 above)
Comment (jnm2) [reply src/NUnitFramework/framework/Internal/Results/TestResult.cs]: I mean... maybe. StringWriter is a minimal wrapper around a StringBuilder. So long as the instance doesn't leave this scope this doesn't bother me either way. Of course then, why not use a StringBuilder directly?

File: src/NUnitFramework/tests/Assertions/AssertFailTests.cs (hunk 12 above)
Comment (CharliePoole) [reply src/NUnitFramework/tests/Assertions/AssertFailTests.cs]: @julealgon Which of the two approaches were you giving the thumbs up to?

File: src/NUnitFramework/tests/Assertions/AssertFailTests.cs (hunk 13 above)
Comment (jnm2) [reply src/NUnitFramework/tests/Assertions/AssertFailTests.cs]: @julealgon I favor your approach because it yields a clearer failure message, but I think this is a matter of style no different than choosing between `collection.Count == 0` and `collection.Any()`.

**Comment by CharliePoole** [comment]
@julealgon Some of your comments are what I'd call "polishing" the code. I like them but I'll wait to see if the overall approach is satisfactory to everyone before spending time. Don't like to "polish" and then discard!

File: src/NUnitFramework/tests/Assertions/AssertFailTests.cs (hunk 13 above)
Comment (CharliePoole) [reply src/NUnitFramework/tests/Assertions/AssertFailTests.cs]: @jnm2 Sold! Based on the failure message.

I still hold back on a lot of Linq stuff for the same reason you like it. It just doesn't speak to me. I have to think twice to remember what each method does, whereas I know immediately what count == 0 means. That's particularly true with a method like Any() - a name that has its own use in NUnit. 

I wish we had a bit more specialization. Then, the people who did most of the work in a particular area of the code could decide on stylistic issues to their liking. Those who only dabbled would use that style for consistency. Of course, I was the guy for a long time in this particular area... 😄

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs (hunk 7 above)
Comment (julealgon) [reply src/NUnitFramework/framework/Internal/Results/TestResult.cs]: I know it's not that big of a deal here, but it is a known good practice to always dispose `IDisposable` instances. There is just no reason not to in this case as well.

I don't mind if you want to change it into a `StringBuilder` though.

File: src/NUnitFramework/tests/Assertions/AssertFailTests.cs (hunk 13 above)
Comment (jnm2) [reply src/NUnitFramework/tests/Assertions/AssertFailTests.cs]: Ironically, contrary to what I just said, I still instinctively use `.Count == 0` rather than `.Any()`. I'm still trying to kick the habit as a premature optimization.

File: src/NUnitFramework/framework/Internal/Results/TestResult.cs (hunk 7 above)
Comment (jnm2) [reply src/NUnitFramework/framework/Internal/Results/TestResult.cs]: The argument in favor of doing this is that StringWriter may change to do something like lease a StringBuilder and return it to the pool on Dispose. Even though I can imagine the framework compat council vetoing that because people have not always disposed it, you are right about this.

File: src/NUnitFramework/tests/Assertions/AssertFailTests.cs (hunk 12 above)
Comment (julealgon) [reply src/NUnitFramework/tests/Assertions/AssertFailTests.cs]: > @julealgon Which of the two approaches were you giving the thumbs up to?

I was just agreeing with your comment in general, in that I understand the huge challenge it would be to do something different right now, even though I'm still not fond of the proposed approach.

As I said, it is infinitely more intuitive to control failures _and_  test termination via the same exceptions, but it looks like the entire team is set on the idea of having these concerns separated. I'm leaving this one up to you guys.

**Comment by CharliePoole** [comment]
I'll have to refresh my memory on this one to see what changes were asked in the reviews. IIRC they were all incidental.

**Comment by rprouse** [comment]
The changes were incidental (that is why I approved), but there were a few that you mentioned were good ideas. Too many comments to read through 😄

**Comment by CharliePoole** [comment]
Rebased on current master.

**Comment by rprouse** [comment]
AppVeyor build failure is one of the unrelated .NET 3.5 failures. I am good to override and merge if you are done @CharliePoole

File: src/NUnitFramework/framework/Exceptions/MultipleAssertException.cs [hunk 16]
```diff
@@ -45,11 +45,15 @@ public class MultipleAssertException : ResultStateException
         /// used normally, when exiting the multiple assert block with failures.
         /// Not used internally but provided to facilitate debugging.
         /// </summary>
-        /// <param name="result">The current result</param>
-        public MultipleAssertException(ITestResult result)
-            : base(result.Message)
+        /// <param name="testResult">
+        /// The current result, up to this point. The result is not used
```
Comment (julealgon) [inline src/NUnitFramework/framework/Exceptions/MultipleAssertException.cs]: Did you forget to remove this from the ctor?

File: src/NUnitFramework/framework/Exceptions/MultipleAssertException.cs [hunk 17]
```diff
@@ -45,11 +45,15 @@ public class MultipleAssertException : ResultStateException
...
-            : base(result.Message)
+        /// <param name="testResult">
+        /// The current result, up to this point. The result is not used 
+        /// internally by NUnit but is provided to facilitate debugging.
+        /// </param>
+        public MultipleAssertException(ITestResult testResult)
+            : base(testResult?.Message)
         {
-            TestResult = result;
+            Guard.ArgumentNotNull(testResult, "testResult");
```
Comment (julealgon) [inline src/NUnitFramework/framework/Exceptions/MultipleAssertException.cs]: Can we use `nameof(testResult)` instead?

Comment (julealgon) [inline src/NUnitFramework/framework/Exceptions/MultipleAssertException.cs]: I'm still not convinced this is a good approach, because argument validation is being performed _after_ calling the base constructor. In these cases, inline static validation like what I posted earlier seems like the better way to go.

Comment (CharliePoole) [reply src/NUnitFramework/framework/Exceptions/MultipleAssertException.cs]: Didn't see this one before merging. I considered creating some new utilities along the line of what you suggested to ensure the arg is not null but decided it was out of scope. The call to the base is made to be OK, but the derived constructor will then throw with a proper message. That seems good enough for now.

File: src/NUnitFramework/tests/Assertions/AssertFailTests.cs (hunk 12 above)
Comment (jnm2) [inline src/NUnitFramework/tests/Assertions/AssertFailTests.cs]: @CharliePoole Since we've been telling folks not to rely on this behavior, shouldn't we switch this from TODO to by-design (remove Ignore and negate the test)?

File: src/NUnitFramework/tests/Assertions/AssertWarnTests.cs [hunk 18]
```diff
@@ -51,5 +52,47 @@ public void AssertWarnWorksWithMessageAndArgs()
...
+            Assert.That(result.Message, Contains.Substring("First warning"));
+            Assert.That(result.Message, Contains.Substring("Second warning"));
+            Assert.That(result.Message, Contains.Substring("This fails"));
+        }
+
+        [Test, Ignore("Currently Fails: Ignored message is displayed without the warnings")]
```
Comment (jnm2) [inline src/NUnitFramework/tests/Assertions/AssertWarnTests.cs]: @CharliePoole Could you start an issue and gather the information into one place so we can understand how to get to work on these new TODO items?

Comment (CharliePoole) [reply src/NUnitFramework/tests/Assertions/AssertWarnTests.cs]: @jnm2 I take it that by TODOs you mean the two ignored tests in this class? Do we know that the tests still don't work?

Comment (jnm2) [reply src/NUnitFramework/tests/Assertions/AssertWarnTests.cs]: They both still fail on master if I unignore them.

Comment (jnm2) [reply src/NUnitFramework/tests/Assertions/AssertWarnTests.cs]: Yes. They both still fail on master if I unignore them.

Comment (jnm2) [reply src/NUnitFramework/tests/Assertions/AssertWarnTests.cs]: @CharliePoole Three TODOs were introduced by this PR. https://github.com/nunit/nunit/pull/2924 handles one which is not related to these two, as far as I can see. I'd appreciate having the information about how to follow up in a new assignable issue, if you're willing and any of this is still fresh in your mind.

File: src/NUnitFramework/tests/Assertions/AssertFailTests.cs (hunk 12 above)
Comment (jnm2) [reply src/NUnitFramework/tests/Assertions/AssertFailTests.cs]: Fixed by https://github.com/nunit/nunit/pull/2924.

File: src/NUnitFramework/tests/Assertions/AssertWarnTests.cs (hunk 18 above)
Comment (CharliePoole) [reply src/NUnitFramework/tests/Assertions/AssertWarnTests.cs]: I'm totally confused now! AFAIK a TODO is a formatted comment that starts `// TODO`, which is specially handled by Visual Studio. I also understand that "todo" is a generic phrase for something that needs to be done. Can you tell me what you mean by it? Where are the three todos? Can I read them?

Earlier in the thread, you asked me to create an issue, which is a term we do have a common understanding of (I think), but now you seem to have resolved the issues in a PR.

I'm almost 100% sure that you are using a term you think I'll understand 😈 but it's not working!

Comment (jnm2) [reply src/NUnitFramework/tests/Assertions/AssertWarnTests.cs]: I'm so sorry! Replace "TODO" with "Ignore" and the above should make sense.
I had internalized that "Ignore" means "use sparsely and over short periods of time," but I hadn't internalized that this was different than a TODO.
